rpmio/rpmpgp.c

```diff
@@ -1057,38 +1055,135 @@ unsigned int pgpDigParamsAlgo(pgpDigParams digp, unsigned int algotype)
+{
+    pgpDigParams digp = xcalloc(1, sizeof(*digp));
+    digp->tag = tag;
+    return digp;
+}
+
+static void hashKey(DIGEST_CTX hash, const struct pgpPkt *pkt)
+{
+    uint8_t head[] = {
+	0x99,
+	(pkt->blen >> 8),
+	(pkt->blen     ),
+    };
+
+    rpmDigestUpdate(hash, head, 3);
+    rpmDigestUpdate(hash, pkt->body, pkt->blen);
+}
+
+static void hashUser(DIGEST_CTX hash, const struct pgpPkt *pkt)
+{
+    uint8_t head[] = {
+	0xb4,
+	(pkt->blen >> 24),
+	(pkt->blen >> 16),
+	(pkt->blen >>  8),
+	(pkt->blen      ),
+    };
+    rpmDigestUpdate(hash, head, 5);
+    rpmDigestUpdate(hash, pkt->body, pkt->blen);
+}
+
+static int pgpVerifySelf(pgpDigParams key, pgpDigParams selfsig,
+			const struct pgpPkt *all, int i)
+{
+    int xx = -1;
+    DIGEST_CTX hash = rpmDigestInit(selfsig->hash_algo, 0);
+    hashKey(hash, &all[0]);
+
+    switch (selfsig->sigtype) {
+    case PGPSIGTYPE_SUBKEY_BINDING:
+	hashKey(hash, &all[i-1]);
+	break;
+    case PGPSIGTYPE_GENERIC_CERT:
+    case PGPSIGTYPE_PERSONA_CERT:
+    case PGPSIGTYPE_CASUAL_CERT:
+    case PGPSIGTYPE_POSITIVE_CERT:
+	hashUser(hash, &all[i-1]);
+	break;
+    default:
+	/* ignore unknown types */
+	xx = 0;
+	break;
+    }
+
+    if (xx)
+	xx = pgpVerifySignature(key, selfsig, hash);
+
+    rpmDigestFinal(hash, NULL, NULL, 0);
+
+    return xx;
+}
+
 int pgpPrtParams(const uint8_t * pkts, size_t pktlen, unsigned int pkttype,
 		 pgpDigParams * ret)
 {
     const uint8_t *p = pkts;
     const uint8_t *pend = pkts + pktlen;
     pgpDigParams digp = NULL;
-    struct pgpPkt pkt;
+    pgpDigParams selfsig = NULL;
+    int i = 0;
+    int alloced = 16; /* plenty for normal cases */
+    struct pgpPkt *all = xmalloc(alloced * sizeof(*all));
     int rc = -1; /* assume failure */
+    int expect = 0;
 
     while (p < pend) {
-	if (decodePkt(p, (pend - p), &pkt))
+	struct pgpPkt *pkt = &all[i];
+	if (decodePkt(p, (pend - p), pkt))
 	    break;
 
+	if (pkt->tag == PGPTAG_PUBLIC_SUBKEY)
+	    expect = PGPTAG_SIGNATURE;
+
 	if (digp == NULL) {
-	    if (pkttype && pkt.tag != pkttype) {
+	    if (pkttype && pkt->tag != pkttype) {
 		break;
 	    } else {
-		digp = xcalloc(1, sizeof(*digp));
-		digp->tag = pkt.tag;
+		digp = pgpDigParamsNew(pkt->tag);
 	    }
 	}
 
-	if (pgpPrtPkt(&pkt, digp))
+	if (digp->tag == PGPTAG_PUBLIC_KEY && pkt->tag == PGPTAG_SIGNATURE)
+	    selfsig = pgpDigParamsNew(pkt->tag);
+
+	if (pgpPrtPkt(pkt, selfsig ? selfsig : digp))
 	    break;
 
-	p += (pkt.body - pkt.head) + pkt.blen;
+	/* subkeys must be followed by binding signature */
+	if (i > 0 && all[i-1].tag == PGPTAG_PUBLIC_SUBKEY)
+	    if (!(selfsig && selfsig->sigtype == PGPSIGTYPE_SUBKEY_BINDING))
+		break;
+
+	if (selfsig) {
+	    int xx = -1;
+
+	    if (i > 0)
+		xx = pgpVerifySelf(digp, selfsig, all, i);
+
+	    selfsig = pgpDigParamsFree(selfsig);
+	    if (xx)
+		break;
+	    expect = 0;
+	}
+
+	i++;
+	p += (pkt->body - pkt->head) + pkt->blen;
 	if (pkttype == PGPTAG_SIGNATURE)
 	    break;
+
+	if (alloced < i) {
```

Comment:
Shouldn't that be `alloced <= i`?

Reply:
:facepalm: 
Thanks for spotting, the '<' is a remnant from a different version where it was the right thing, but wrong here of course.